File: ServiceTest4/app/src/main/java/activitytest/example/com/servicetest4/MainActivity.java
package activitytest.example.com.servicetest4;

import android.content.ComponentName;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.IBinder;
import android.os.PersistableBundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;

public class MainActivity extends AppCompatActivity implements View.OnClickListener{

    private MyService.DownloaderBinder mDownloaderBinder;

    private ServiceConnection mConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            mDownloaderBinder = (MyService.DownloaderBinder) service;
            mDownloaderBinder.startDownload();
            mDownloaderBinder.getProgress();
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {

        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Button startService = findViewById(R.id.start_service);
        Button stopService = findViewById(R.id.stop_service);

        startService.setOnClickListener(this);
        stopService.setOnClickListener(this);

        Button bindService = findViewById(R.id.start_service);
        Button unbindService = findViewById(R.id.unbind_service);

        bindService.setOnClickListener(this);
        unbindService.setOnClickListener(this);


    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.start_service:
                Intent startIntent = new Intent(this,MyService.class);
                startService(startIntent);
                break;
            case R.id.stop_service:
                Intent stopIntent = new Intent(this,MyService.class);
                stopService(stopIntent);
                break;
            case R.id.bind_service:
                Intent bindIntent = new Intent(this,MyService.class);
                bindService(bindIntent,mConnection,BIND_AUTO_CREATE);
                break;
            case R.id.unbind_service:
                unbindService(mConnection);
                break;
            default:
                break;

        }
    }
}

File: ServiceTest3/app/src/main/java/com/example/servicetest/MainActivity.kt
package com.example.servicetest

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.IBinder
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.widget.Button

class MainActivity : AppCompatActivity(), View.OnClickListener {

    private var downloadBinder: MyService.DownloadBinder? = null
    private val connection = object : ServiceConnection {

        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            downloadBinder = service as MyService.DownloadBinder
            downloadBinder!!.startDownload()
            downloadBinder!!.progress
        }

        override fun onServiceDisconnected(name: ComponentName) {

        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val startService = findViewById(R.id.start_service) as Button
        val stopService = findViewById(R.id.stop_service) as Button
        startService.setOnClickListener(this)
        stopService.setOnClickListener(this)
        val bindService = findViewById(R.id.bind_service) as Button
        val unbindService = findViewById(R.id.unbind_servce) as Button
        bindService.setOnClickListener(this)
        unbindService.setOnClickListener(this)
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.start_service -> {
                val startIntent = Intent(this, MyService::class.java)
                startService(startIntent)
            }
            R.id.stop_service -> {
                val stopIntent = Intent(this, MyService::class.java)
                stopService(stopIntent)
            }
            R.id.bind_service -> {
                val bindIntent = Intent(this, MyService::class.java)
                bindService(bindIntent, connection, Context.BIND_AUTO_CREATE)
            }
            R.id.unbind_servce -> unbindService(connection)//解绑服务
            else -> {
            }
        }
    }
}







